README.md
[![Moleculer](https://badgen.net/badge/Powered%20by/Moleculer/0e83cd)](https://moleculer.services)

# newFlowTS

## Informações Tecnicas

-   [Moleculer](https://moleculer.services/)

## Boas Práticas

Segue abaixo os padrões acordados pelo time de Integração:

-   Escrita de função e arquivos utilizando [CamelCase](https://pt.wikipedia.org/wiki/CamelCase)
-   Utilizar padrão declarativo de variável do TypeScript. exemplo? var x srá const ou let x.
-   Não alterar o padrão de identação sem aviso prévio
-

OBS. Desenvolvimento fora do padrão acima terá as PRs reprovadas automaticamente.

## Cuidados

Por se tratar de um código que terá chamados sincronas e assincronas, se faz necessários o entendimento dos processos envolvidos para realização da transação. Em casos de duvidas procurar algum membro do time de integração.

## Uso

Inicie o projeto com o comando `npm run dev`.
Após iniciar, abra a URL http://localhost:3000/ em seu navegador.
Na página de boas-vindas, você pode testar os serviços gerados via API Gateway e verificar os nós e serviços.

No terminal, tente os seguintes comandos:

-   `nodes` - Liste todos os nós conectados.
-   `actions` - Liste todas as ações de serviço registradas.

## Services

-   **api**: API Gateway services

## NPM scripts

-   `npm run dev`: Inicie o modo de desenvolvimento (carregue todos os serviços localmente com hot-reload e REPL)
-   `npm run start`: inicia o modo de produção (defina a variável env `SERVICES` para carregar determinados serviços)
-   `npm run cli`: Inicie uma CLI e conecte-se à produção. Não se esqueça de definir o namespace de produção com o argumento `--ns` no script
-   `npm run lint`: executa o ESLint
-   `npm run ci`: execute o modo de teste contínuo com observação
-   `npm test`: execute testes e gere relatório de cobertura
-   `npm run dc:up`: Inicie a pilha com o Docker Compose
-   `npm run dc:down`: pare a pilha com o Docker Compose

## Arquitetura

-   Nossa estrutura foi definida seguindo as recomendações do moleculer:
    -   service ou src - camada de programa ==> nome do programa que conterá os serviços (exemplo: crmIso, erpProtheusVivo, etc) - Dominio ==> nome do domino de execução que o serviço se enquadra(order, invoice, consume, etc)
        service ==> destinada ao gerenciamento dos serviços, todos os arquivos dentro dessa pasta deverão ser "service.ts"
        src ==> nessa pasta será possivel encontrar a pasta entity, repository e interfaces.

Qualquer duvida procurar os responsáveis

## Não Commitar

package-lock.json
package.json - Apenas se houver nova dependência
.vscode

## Acessando o OPENAPI

{url}/api/openapi

## Camadas

integration-layer - Midiaware centralizador das integrações
crm-iso - integração do sistema cremmer com protheus
wms-alcis - integração do wms da alcis com protheus
erp-protheus-viveo - gateway centralizador das integraçõs

## Padrões

-   Services:
    `module name` (pasta) - camelCase descritivo base do que é o sistema (wms, erp, crm ou camada) + nome do sistema ex: crmIso.
    `sub module` - no do serviço sempre no singular ex: order, payment e etc.
    `name arquive` - camelCase
    `classes` - No padrão Typescript
    `variables` - No padrão Typescript (const e let)
    `cronJob` - Usando a library CronJob

-   Entity:
    `views (ViewEntity)` - nome camelCase começando com vw minusculo
    `tabelas (Entity)` - nome camelCase começando com tb minusculo

FLOW CREMER


CHANGELOG.md
# Changelog

Notação das versões e implementações realizadas por Versão

O formato de notação tomou como base [Utilizando o Changelog](https://keepachangelog.com/en/1.0.0/)
e para determinar o versionamento foi usado [Versionamento Semantico](https://semver.org/lang/pt-BR/)

O Versionamento foi iniciado a partir da versão 1.0.0 publicada até o dia 31/12/2022

<<<<<<< HEAD
## [x.xx.xx] - xx/xx/xxxx

-   Camada de Integração
    -   Implementado teste unitários com cobertura minima de 30%
    -   Teste iniciados pela pasta service e suas subpastas
    -   Configuração do JEST extraída do arquivo package.json e implementada no arquivo jest.config.ts
    -   Atualização do arquivo GITIGNORE

## [1.32.0] - 21/09/2023
=======
## [1.34.0] - 21/09/2023

### Added

    - Integração Alcis - Protheus
        - Incluído nova filial Catalão M03.

## [1.33.0] - 21/09/2023
>>>>>>> origin/master

### Added

    - Integração ExpedLOG - Devolução
        - Incluído endpoints para buscar e integrar dados de devolução.

## [1.32.0] - 21/09/2023

### Added

-   Integração Total Express
    -   Envio das Notas Fiscais via API para a TOTAL express controlando o envio das mesmas via banco, garantindo o envio de todas as notas fiscais

## [1.31.0] - 13/09/2023

### Changed

-   Integração Ecommerce
    -   Alteração busca de produtos para receber a aliquota de IPI e informação se produto é importado, armazenamento da informação em banco de dados e envio ao Ecommerce
    -   Alterado inserção de pedido no protheus de origem do Ecommerce, para verificar se produto a ser inserido é importado ou não, caso importado é realizado o calculo para saber o valor da aliquito do IPI e deduzido do valor bruto para ser enviado ao PROTHEUS
    -   CRONs transformados em variaveis ENV para manter o padrão a camada.

## [1.30.3] - 06/09/2023

### Changed

-   Integração WMS Alcis

    -   Reprocessar confirmacao de pedido quando retornado status 500 do Protheus

## [1.30.2] - 05/09/2023

### Changed

-   Integração WMS Alcis

    -   Incluído Cremer no Order Confirmation

## [1.30.1] - 05/09/2023

### Changed

-   Integração WMS Alcis

    -   Alterado Cron Job Order Confirmation

## [1.30.0] - 05/09/2023

### Added

-   Integração WMS Alcis

    -   Melhorias na integração de pedidos e notas
    -   Camada recebe os pedidos e processa o envio para o ERP Protheus
    -   Controle de Status

## [1.29.7] - 01/09/2023

### Changed

-   Integração WMS Alcis V1.3
    -   Ajustes Axios retorno
    -   Melhoria das validações de retorno em que o Protheus devolve string no status.
    -   Melhoria nos methodos.

## [1.29.6] - 31/08/2023

### Changed

-   Integração WMS Alcis V1.3 - Correção axios se perdendo nos requests. - Melhoria das validações. - Criação de travas por filial

## [1.29.5] - 30/08/2023

### Changed

-   Integração WMS Alcis V1.3

    -   Acrescimo de index e logs especificos na confirmação de pedido.
    -   Retorno do uso do broker para balanceamento de ambas rotas.

## [1.29.4] - 29/08/2023

### Changed

-   Integração WMS Alcis V1.3

    -   Segregação execução Axios Filiais

## [1.29.3] - 29/08/2023

### Changed

-   Integração WMS Alcis V1.3

    -   Melhoria rota pararelização

## [1.29.2] - 29/08/2023

### Changed

-   Integração WMS Alcis V1.3

    -   Trava de instancias do CroJob
    -   Melhorias nos metodos de paralelização

## [1.29.1] - 25/08/2023

### Changed

-   Integração WMS Alcis V1.2
    -   Ajustes parametrização cronjob e flag

## [1.29.0] - 25/08/2023

### Changed

-   Integração WMS Alcis V1.2
    -   Melhorias na integração de reservas - I2308-10405
    -   Camada recebe os pedidos e processa o envio para o ERP Protheus
    -   Controle de Status

## [1.28.1] - 23/08/2023

### Changed

-   Novo Flow Cremer
    -   Validação de status do Pedido no protheus, capturado pelo número CRM, para verificar se o status do pedido está cancelado.

## [1.28.0] - 21/08/2023

### Changed

-   Alteração URL Integração Ouroweb
    -   Realizada a alteração da URL da Ouro, por conta da migração do ERP para nuvem.

## [1.27.0] - 21/08/2023

### Changed

-   Ajuste Cors
    -   Foi realizado o ajuste do Cors, alterando para URL de produção. Essa alteração foi feita para sanar uma vulnerabilidade apontada pelo Sonar Qualys e solicitado pelo time de segurança Viveo.

## [1.26.0] - 21/08/2023

### Changed

-   Novo Flow Cremer

    -   Ajuste no processo de geração da Ordem para o Protheus
    -   Incluíndo retry para busca de dados do processo do flow

-   Atualização de Libraries

    -   Foi realizado a atualizaçao do pacote elastic-apm-node para a versão 3.49 para resolver 2 vulnerabilidades encontrada no pacote import-in-the-middle utilizado pelo primeiro pacote citado.

-   Ecommerce
    -   Atualização do startTransaction do Ecommerce, melhorando a nomeclatura do texto

## [1.25.0] - 21/08/2023

### Added

    - Integração Alcis/Protheus
        - Alteração do REST Protheus para utilização de Broker

## [1.23.1] - 16/08/2023

### Fixed

-   Correção
    -   Correção CronJob Iso Crm

## [1.23.0] - 16/08/2023

### Changed

-   Integração Alcis
    -   Novo sistema token Camada-Alcis, bem mais performático e com taxa de erro sob "stress test" próximo de nulo.
    -   Metricas por service no APM, loggando cada service name

## [1.22.0] - 14/08/2023

### Changed

-   Novo Flow Cremer
    -   Ajustes no busca de pagamentos do ISO CRM sem pedido GERADO para gravar o dados vazio
    -   Ajustes retry do pedidos com status aguardando processamento para voltar ao status Recebido Integrador
    -   Gravação de orders_checks na geração de OF e gravação de pagamento

## [1.21.0] - 11/08/2023

### Changed

-   Novo Flow Cremer

    -   O For utilizado para extrair as informações de pagamento foram ajustadas para utilização do FOR OF ao inves do FOR LOOP
    -   Aplicado ajustes no processo de manutenção das OFs para remover a duplicidade de dados.
    -   Incluido a inserção dos dados sobre pagamento na orders_checks

moleculer.config.ts
'use strict';
import { BrokerOptions, Errors, MetricRegistry } from 'moleculer';
import { formatToTimeZone } from 'date-fns-timezone';

/**
 * Moleculer ServiceBroker configuration file
 *
 * More info about options:
 *     https://moleculer.services/docs/0.14/configuration.html
 *
 *
 * Overwriting options in production:
 * ================================
 *    You can overwrite any option with environment variables.
 *    For example to overwrite the "logLevel" value, use `LOGLEVEL=warn` env var.
 *    To overwrite a nested parameter, e.g. retryPolicy.retries, use `RETRYPOLICY_RETRIES=10` env var.
 *
 *    To overwrite broker’s deeply nested default options, which are not presented in "moleculer.config.js",
 *    use the `MOL_` prefix and double underscore `__` for nested properties in .env file.
 *    For example, to set the cacher prefix to `MYCACHE`, you should declare an env var as `MOL_CACHER__OPTIONS__PREFIX=mycache`.
 *  It will set this:
 *  {
 *    cacher: {
 *      options: {
 *        prefix: "mycache"
 *      }
 *    }
 *  }
 */
const brokerConfig: BrokerOptions = {
	// Namespace of nodes to segment your nodes on the same network.
	namespace: 'flow',
	// Unique node identifier. Must be unique in a namespace.
	nodeID: 'flow',
	// Custom metadata store. Store here what you want. Accessing: `this.broker.metadata`
	metadata: {},

	// Enable/disable logging or use custom logger. More info: https://moleculer.services/docs/0.14/logging.html
	// Available logger types: "Console", "File", "Pino", "Winston", "Bunyan", "debug", "Log4js", "Datadog"
	logger: {
		type: 'Console',
		options: {
			// Using colors on the output
			colors: true,
			// Print module names with different colors (like docker-compose for containers)
			moduleColors: true,
			// Line formatter. It can be "json", "short", "simple", "full", a `Function` or a template string like "{timestamp} {level} {nodeID}/{mod}: {msg}"
			formatter: `${formatToTimeZone(new Date(), 'YYYY-MM-DD HH:mm:ss', {
				timeZone: 'America/Sao_Paulo'
			})} {level} {nodeID}/{mod}: {msg}`,
			// Custom object printer. If not defined, it uses the `util.inspect` method.
			objectPrinter: null,
			// Auto-padding the module name in order to messages begin at the same column.
			autoPadding: false
		}
	},
	// Default log level for built-in console logger. It can be overwritten in logger options above.
	// Available values: trace, debug, info, warn, error, fatal
	logLevel: 'info',

	// Define transporter.
	// More info: https://moleculer.services/docs/0.14/networking.html
	// Note: During the development, you don't need to define it because all services will be loaded locally.
	// In production you can set it via `TRANSPORTER=nats://localhost:4222` environment variable.
	transporter: null, // "TCP"

	// Define a cacher.
	// More info: https://moleculer.services/docs/0.14/caching.html
	cacher: null,

	// Define a serializer.
	// Available values: "JSON", "Avro", "ProtoBuf", "MsgPack", "Notepack", "Thrift".
	// More info: https://moleculer.services/docs/0.14/networking.html#Serialization
	serializer: 'JSON',

	// Number of milliseconds to wait before reject a request with a RequestTimeout error. Disabled: 0
	requestTimeout: 10 * 60000,

	// Retry policy settings. More info: https://moleculer.services/docs/0.14/fault-tolerance.html#Retry
	retryPolicy: {
		// Enable feature
		enabled: false,
		// Count of retries
		retries: 5,
		// First delay in milliseconds.
		delay: 100,
		// Maximum delay in milliseconds.
		maxDelay: 1000,
		// Backoff factor for delay. 2 means exponential backoff.
		factor: 2,
		// A function to check failed requests.
		check: (err: Errors.MoleculerError) => err && !!err.retryable
	},

	// Limit of calling level. If it reaches the limit, broker will throw an MaxCallLevelError error. (Infinite loop protection)
	maxCallLevel: 100,

	// Number of seconds to send heartbeat packet to other nodes.
	heartbeatInterval: 10,
	// Number of seconds to wait before setting node to unavailable status.
	heartbeatTimeout: 30,

	// Cloning the params of context if enabled. High performance impact, use it with caution!
	contextParamsCloning: false,

	// Tracking requests and waiting for running requests before shuting down. More info: https://moleculer.services/docs/0.14/context.html#Context-tracking
	tracking: {
		// Enable feature
		enabled: false,
		// Number of milliseconds to wait before shuting down the process.
		shutdownTimeout: 5000
	},

	// Disable built-in request & emit balancer. (Transporter must support it, as well.). More info: https://moleculer.services/docs/0.14/networking.html#Disabled-balancer
	disableBalancer: false,

	// Settings of Service Registry. More info: https://moleculer.services/docs/0.14/registry.html
	registry: {
		// Define balancing strategy. More info: https://moleculer.services/docs/0.14/balancing.html
		// Available values: "RoundRobin", "Random", "CpuUsage", "Latency", "Shard"
		strategy: 'RoundRobin',
		// Enable local action call preferring. Always call the local action instance if available.
		preferLocal: true
	},

	// Settings of Circuit Breaker. More info: https://moleculer.services/docs/0.14/fault-tolerance.html#Circuit-Breaker
	circuitBreaker: {
		// Enable feature
		enabled: false,
		// Threshold value. 0.5 means that 50% should be failed for tripping.
		threshold: 0.5,
		// Minimum request count. Below it, CB does not trip.
		minRequestCount: 20,
		// Number of seconds for time window.
		windowTime: 60,
		// Number of milliseconds to switch from open to half-open state
		halfOpenTime: 10 * 1000,
		// A function to check failed requests.
		check: (err: Errors.MoleculerError) => err && err.code >= 500
	},

	// Settings of bulkhead feature. More info: https://moleculer.services/docs/0.14/fault-tolerance.html#Bulkhead
	bulkhead: {
		// Enable feature.
		enabled: false,
		// Maximum concurrent executions.
		concurrency: 20,
		// Maximum size of queue
		maxQueueSize: 150
	},

	// Enable action & event parameter validation. More info: https://moleculer.services/docs/0.14/validating.html
	validator: true,

	errorHandler: null,

	// Enable/disable built-in metrics function. More info: https://moleculer.services/docs/0.14/metrics.html
	metrics: {
		enabled: true,
		// Available built-in reporters: "Console", "CSV", "Event", "Prometheus", "Datadog", "StatsD"
		reporter: {
			type: 'Prometheus',
			options: {
				// HTTP port
				port: 3039,
				// HTTP URL path
				path: '/metrics',
				// Default labels which are appended to all metrics labels
				defaultLabels: (registry: MetricRegistry) => ({
					namespace: registry.broker.namespace,
					nodeID: registry.broker.nodeID
				})
			}
		}
	},

	// Enable built-in tracing function. More info: https://moleculer.services/docs/0.14/tracing.html
	tracing: {
		enabled: true,
		// Available built-in exporters: "Console", "Datadog", "Event", "EventLegacy", "Jaeger", "Zipkin"
		exporter: {
			type: 'Jaeger', // Console exporter is only for development!
			options: {
				// HTTP Reporter endpoint. If set, HTTP Reporter will be used.
				endpoint: null,
				// UDP Sender host option.
				host: '127.0.0.1',
				// UDP Sender port option.
				port: 6832,
				// Jaeger Sampler configuration.
				sampler: {
					// Sampler type. More info: https://www.jaegertracing.io/docs/1.14/sampling/#client-sampling-configuration
					type: 'Const',
					// Sampler specific options.
					options: {}
				},
				// Additional options for `Jaeger.Tracer`
				tracerOptions: {},
				// Default tags. They will be added into all span tags.
				defaultTags: null
			}
		}
	},

	// Register custom middlewares
	middlewares: []

	// Register custom REPL commands.
	//replCommands: null,
	/*
	  // Called after broker created.
	  created : (broker: ServiceBroker): void => {},
	  // Called after broker started.
	  started: async (broker: ServiceBroker): Promise<void> => {},
	  stopped: async (broker: ServiceBroker): Promise<void> => {},
	   */
};

export = brokerConfig;
